api/agent/agent_launcher.go

```diff
@@ -0,0 +1,150 @@
+package agent
+
+import (
+	"context"
+	"time"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+type WaitResult struct {
+	Slot Slot
+	Err  error
+}
+
+// tryGetToken attempts to fetch/acquire a token from slot queue without blocking.
+func tryGetToken(ch chan WaitResult) (Slot, error) {
+	select {
+	case s := <-ch:
+		return s.Slot, s.Err
+	default:
+		return nil, nil
+	}
+}
+
+// waitGetToken blocks and waits on both wait and token channels.
+func waitGetToken(ch chan WaitResult, wait chan struct{}) (Slot, error) {
+	select {
+	case s := <-ch:
+		return s.Slot, s.Err
+	case <-wait:
+	}
+	return nil, nil
+}
+
+// listenWaitResult listens [context, agent shutdown, slot channel] and pushes the result to one channel
+func (a *agent) listenWaitResult(ctx context.Context, call *call, slotChan chan *slotToken) chan WaitResult {
+	output := make(chan WaitResult)
+	dampen := make(chan struct{}, 1)
+
+	go func() {
+		dampen <- struct{}{}
+		for {
+			select {
+			case s := <-slotChan:
+				if call.slots.acquireSlot(s) {
+					if s.slot.Error() != nil {
+						s.slot.Close()
+						output <- WaitResult{nil, s.slot.Error()}
```

Comment:
this needs to be in a select, as well. (i don't think this WaitResult makes life any easier, the slot includes an error too already we could use when we need to generate one from this level, but that only seems like for server shutdown, which we really don't need to do anything except bail and never send anything if we want)

Reply:
i think it gives us some sugar for checking if we acquire the slot right?

Reply:
yes, so this is to make slot/acquire thing easier. The way it is right now is that if acquire fails, you have to retry. But this causes, extra for-loop, so I wrapped all this into this WaitResult channel... It includes ctx and shutdown, so caller checkLaunch becomes simpler...